torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1372,6 +1372,106 @@ std::vector<For*> LoopNest::distributeLoopOverInnerLoops(For* loop) {
   return distributeLoop(loop, loopsSet);
 }
 
+For* LoopNest::fuseLoops(const std::vector<For*>& loops) {
+  if (loops.empty()) {
+    return nullptr;
+  }
+  if (loops.size() == 1) {
+    return loops.front();
+  }
+
+  // Check if all the loops have the same parent.
+  auto root = loops.front()->get_parent();
+  for (auto l : loops) {
+    auto par = l->get_parent();
+    if (par == nullptr) {
+      throw malformed_input("Loop without parent: ", l);
+    }
+    if (par != root) {
+      throw malformed_input("Can't fuse loops with different parents: ", l);
+    }
+  }
+  auto root_block = dynamic_cast<Block*>(root);
+  if (root_block == nullptr) {
+    throw malformed_input(
+        "Loops' parent must be a Block, instead found ", root);
+  }
+
+  // Currently, we only handle cases where there are no statements between
+  // the given loops in their parents body. We can possibly relax this
+  // constraint by allowing statements that do not affect the loops being
+  // fused by performing some dependency analysis. TODO.
+  auto it = root_block->begin();
+  for (; it != root_block->end(); ++it) {
+    if (*it == loops.front())
+      break;
+  }
+  TORCH_INTERNAL_ASSERT(it != root_block->end());
+  for (auto l : loops) {
+    if (*it != l) {
+      throw malformed_input(
+          "Only contiguous loops can be fused, found another stmt: ", *it);
+    }
+    ++it;
+  }
+
+  auto first_loop = loops.front();
+
+  // Check if bounds are the same for all the loops.
+  // TODO: The following check does not consider different expressions that
+  // evaluate to the same value as the same. Improve this by performing
+  // expression equality.
+  auto are_bounds_equal = [](const Expr* bound1, const Expr* bound2) {
+    if (bound1->isConstant() && bound2->isConstant()) {
+      return immediateAs<int>(bound1) == immediateAs<int>(bound2);
+    }
```

Comment:
It might be better to compare `IRSimplifier::simplify(bound2-bound1) == IntImm(0)` - this way it would also work for non-constant start/end bounds.

Reply:
I tried this but this seems to give incorrect results. I will debug this more and fix it as a follow up. There is a TODO above for this.